### DISCUSSION TITLE
stg bottom

### DISCUSSION THREAD

**Discussion by clillies** [stacked-git/stgit#454] [Ideas]
When scripting, I often want to find the first commit in my series.
The sha of the first commit can be deduced by other means of course, but I noticed
that the there is a `stg top` command which prints the patch name of the top _applied_ patch.

So, even though it would be nice and probably fairly easy to add `stg bottom`, which would show
the bottom applied patch (or error if none is applied), I see some _overlap_ between
`stg top` and `stg series` and even `stg show`.

Could we instead extend `stg series` and add `--top` and `--bottom` (or perhaps `--only-top` or `--only-bottom`), and some kind of formatter args, `--name-only` to get only the name. The `--top` and `--bottom` would then be affected by `-A` and `-U` and `-H`, instead of just work with `-A` as today.
Then `stg top` could be thanked, pensioned and deprecated/retired, and we would get (at least eventually) one command less, but more functionality in existing commands.

Just an idea.

**Comment by jpgrayson** [comment]
You can use `stg series` the `-P`/`--no-prefix` option and an absolute index of 0 to print the bottom patch. I.e.:

`stg series -P 0`

The topmost patch could similarly be printed by using `@` or the relative index `+0` or `-0`:

`stg series -P @`

So while not as on the nose as having `--top` and `--bottom` options, `stg series` can already largely do what you're requesting.

Note that to see the first unapplied patch (which you might call the bottom of the unapplied), use the `+1` relative index.

`stg series -P +1`

And the last (topmost) unapplied patch can be accessed using `^` (may need to quote that depending on your shell).

`stg series -P '^'`

It would be straightforward to add aliases for one or more of these. E.g.

`git config --global stgit.alias.bottom 'series --no-prefix 0'`

You can then just invoke:

`stg bottom`

I don't think StGit needs more options or commands for printing patches. All the fancy patch locators were added to solve this problem. I could perhaps be convinced about `stg bottom` to have symmetry with `stg top`.

Perhaps its also worth noting that you perhaps don't even really want to print the, e.g. bottom most patch, but instead want to point various other `stg` commands to the bottom patch. Well, good news: that fancy patch locator syntax is available across all commands already. E.g. to refresh changes to the bottom most patch:

`stg refresh -p 0`

**Reply by jpgrayson** [reply]
The [stg(1)](https://stacked-git.github.io/man/stg/) man page covers the patch locator syntax in detail.

**Reply by clillies** [reply]
Thank you! This looks good.
